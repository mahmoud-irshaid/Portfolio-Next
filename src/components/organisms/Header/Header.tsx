'use client';

import { BurgerIcon } from '@components/atoms/BurgerIcon';
import { Container } from '@components/atoms/Container';
import { Logo } from '@components/atoms/Logo';
import { NavigationItem } from '@components/atoms/NavigationItem';
import { ThemeToggle } from '@components/atoms/ThemeToggle';
import { MobileMenu } from '@components/molecules/MobileMenu';

import { useOnKeyDown } from '@hooks/useOnKeyDown';

import { usePathname } from 'next/navigation';
import { useEffect, useState } from 'react';
import Headroom from 'react-headroom';

export const navItems = [
	{
		href: '/',
		title: 'Home',
	},
	{
		href: '/about',
		title: 'About',
	},
	{
		href: '/projects',
		title: 'Projects',
	},
];

export const Header = () => {
	const [isOpen, setIsOpen] = useState(false);
	const pathname = usePathname();

	useEffect(() => {
		if (isOpen) {
			document.body.style.overflow = 'hidden';
		}

		return () => {
			document.body.style.overflow = 'unset';
		};
	}, [isOpen]);

	useEffect(() => {
		setIsOpen(false);
	}, [pathname]);

	useOnKeyDown('Escape', () => setIsOpen(false));

	const navigationVariants = {
		hidden: { opacity: 0, y: -10 },
		visible: (custom: number) => ({
			opacity: 1,
			y: 0,
			transition: { delay: custom },
		}),
	};

	return (
		<Headroom style={{ zIndex: 50 }}>
			<header className="relative z-50 py-8 backdrop-blur-md">
				<Container>
					<div className="flex items-center justify-between">
						<Logo />
						<nav className="hidden md:block">
							<ul className="flex gap-8 text-lg">
								{navItems.map(({ href, title }, i) => (
									<NavigationItem
										href={href}
										title={title}
										key={href}
										variants={navigationVariants}
										initial="hidden"
										animate="visible"
										customDelay={(i + 1) * 0.1}
									/>
								))}
							</ul>
						</nav>
						<button
							className="absolute right-4 top-8 z-50 md:hidden"
							onClick={() => setIsOpen((prev) => !prev)}
							aria-label="Menu"
						>
							<BurgerIcon isOpen={isOpen} />
						</button>
						<div className="hidden md:block">
							<ThemeToggle />
						</div>
					</div>
				</Container>
			</header>
			<MobileMenu isOpen={isOpen} />
		</Headroom>
	);
};
