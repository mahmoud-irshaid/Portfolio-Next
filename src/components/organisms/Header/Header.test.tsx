import { Header } from '../Header';

import { render } from '@testing-library/react';

jest.mock('@components/atoms/NavigationItem');
jest.mock('next/router', () => ({
	__esModule: true,
	useRouter: jest.fn().mockReturnValue({
		route: '',
		pathname: '',
		query: '',
		asPath: '',
	}),
}));
jest.mock('@components/atoms/Logo');

describe('Header', () => {
	it('renders correctly', () => {
		const { container } = render(<Header />);
		expect(container).toMatchSnapshot();
	});
});
