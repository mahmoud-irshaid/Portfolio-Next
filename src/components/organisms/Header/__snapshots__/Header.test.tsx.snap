// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<div>
  <div
    class="headroom-wrapper"
  >
    <div
      class="headroom headroom--unfixed"
      style="position: relative; top: 0px; left: 0px; right: 0px; z-index: 50; transform: translate3D(0, 0, 0);"
    >
      <header
        class="relative z-50 py-8 backdrop-blur-md"
      >
        <div
          class="mx-auto w-full px-4 md:max-w-4xl lg:max-w-7xl"
        >
          <div
            class="flex items-center justify-between"
          >
            <nav
              class="hidden md:block"
            >
              <ul
                class="flex gap-8 text-lg"
              />
            </nav>
            <button
              aria-label="Menu"
              class="absolute right-4 top-8 z-50 md:hidden"
            >
              <div
                class="burger-icon"
              >
                <span />
                <span />
                <span />
                <span />
              </div>
            </button>
            <div
              class="hidden md:block"
            >
              <button
                aria-label="Toggle Dark Mode"
                class="flex items-center justify-center rounded-full border-2 border-grey-300 bg-transparent p-2 ring-grey-300 transition-all hover:ring-2 hover:ring-offset-2 dark:border-grey-700 dark:bg-transparent dark:ring-grey-200 dark:hover:ring-2 dark:hover:ring-offset-2"
                type="button"
              >
                <svg
                  class="h-5 w-5 text-grey-800 dark:text-grey-200"
                  fill="none"
                  stroke="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
                <span
                  class="md:hidden"
                >
                  Switch to 
                  dark
                   mode
                </span>
              </button>
            </div>
          </div>
        </div>
      </header>
    </div>
  </div>
</div>
`;
